oak_runtime/build.rs

```diff
@@ -37,4 +38,16 @@ fn main() {
             ..Default::default()
         },
     );
+
+    // Build assets for the browser client of the introspection server
+    #[cfg(feature = "oak_debug")]
+    Command::new("npm")
```

Comment:
I don't think this is the right place to do this (e.g. this means that running the Rust unit tests will require an npm install.  Maybe a new sub-step in `runner` would work better?

Reply:
I'll check that out :). Build systems 😅

Reply:
Hmm, I suspect that shifting to use `include_bytes!` pretty much requires `build.rs` integration, otherwise a `cargo test` is going to fail to compile because the generated files under `introspection_browser_client/dist/` aren't guaranteed to be available.

(Personally, I think `include_bytes` is going in the wrong direction &ndash; I think the introspection client code should more loosely bound to the Runtime code, not more tightly bound.)

Reply:
Yes, you're correct about that. The only way to use `include_bytes` without `build.rs` integration would be to use build scripts to create a dummy file for tests etc, which admittedly isn't super nice. 

Something along these lines: https://stackoverflow.com/questions/56014764/falling-back-to-alternative-value-if-include-bytes-target-is-missing

Reply:
I think `include_bytes` makes sense because, without that, things would still fail, just at runtime instead of compile time.

The assets need to be built at some point in order for things to work, so may as well fail early if they cannot be built for any reason.

That said, @daviddrysdale I think you are right that a simple `cargo test` would fail if the assets are not available at that time. I still don't feel that requires integrating with build.rs (which I agree would be too much coupling). But I think there are two possible cases:

- the tests actually require the assets to be there, in which case they need to be built anyways; in this case I think we can still assume that the correct steps were run by `runner` prior to the `cargo test` invocation
- the tests do not require the assets, in which case we should consider either making the entire introspection server optional (I don't remember whether it is already controlled by a compile-time feature flag), or have a separate compile-time flag that replaces the `include_bytes` calls with empty vecs, or some other mock assets (e.g. an inline HTML page with an error)

If we allow the lookup at runtime, I am worried things will be more unpredictable and harder to test, since there is so much dependency on the actual files being in the right location and with the right permissions etc, plus associated risks with the server being able to look up arbitrary files etc.

Having a self-contained binary makes a lot of these concerns go away (for instance it would make a hypothetical security review much smoother, in my mind).

Anyways, if @daviddrysdale feels strongly about this, I am happy to go with the previous approach. What's your opinion @juliettepretot ?

Reply:
@tiziano88 Thanks for your reply! 

To answer your question about `cargo test, I tried it, it does fail. They don't technically require the files, but `include_bytes!` calls will fail. Alongside with building the docs and generating test coverage. (I remedied that with creating placeholder files in `build.rs`, but it's [not pretty](https://github.com/project-oak/oak/pull/1382/commits/297c55dff074def0ac4caaac04ad864ea5d479ac). If there's a way to do that differently [?] that'd be cool.)

I share you concerns about serving files from the file system, both for complexity & security concerns. 

While I'm quite happy to bundle the introspection server with the runtime binary, I think there is a third solution, as mentioned [my comment in the other thread](https://github.com/project-oak/oak/pull/1382#discussion_r477233839):

We could move the dynamic introspection frontend logic to a different package, away from the Oak Runtime. The runtime aux server would still provide an API that serves a list of events, but it would be up to the developer to start a different server that reads that API and produces useful output from it. - It'd cleanly separates the two, but I'd caveat that it arguably duplicates some of the aux server logic.

Reply:
That's an interesting suggestion; IIUC, you are saying that we could leave only the API endpoint in the server, but serve static assets via some other mechanism (a la `python3 -m http.server`)? That does seem nicer in terms of separating the two build processes, but it still feels like a lot of moving parts to coordinate, that need to be aligned in order for the final introspection to work.

Reply:
> IIUC, you are saying that we could leave only the API endpoint in the server, but serve static assets via some other mechanism (a la python3 -m http.server)?

Yup, that's my suggestion. :) 

> but it still feels like a lot of moving parts to coordinate, that need to be aligned in order for the final introspection to work

Agreed, which is why I'd also be okay to bundle the two via `include_bytes!`, but no strong opinion. Two bundle the two we'd either a) have to ask developer to themselves start the introspection server in another process or b) start that process via the runner I guess? In the case of b) however we're sort of replicating what an aux server in the runtime is, just in the runner. Eg a server that should start with the runtime, and shut down with the runtime.

Reply:
What do you think cc @daviddrysdale?

Reply:
BTW if tests are the only outstanding issue, it may be possible to stub out the `include_bytes` call like

```rust
if cfg!(test) {
  "test"
else {
  include_bytes!(...)
}
```

WDYT?

Reply:
I personally prefer the idea of having a separate server that serves the static assets, which in this case just happens to be the results of a build of the client-side introspection code, although as you point out it does come with the downside of fiddling with parallel processes in deployment (plus any CORS concerns).

But I think it's worth it for the clear separation between interface and implementation.  The defined interface point is the `/introspection-events` URL serving serialized `Events` proto data, exposed by the Runtime; beyond this there is complete freedom as to what consumes the data from this interface.

The usual consumer will be client-side JS code built from the code in this PR, served by an arbitrary webserver that serves static data (maybe behind a common reverse proxy in a "real-world" deployment?), but that's not absolutely required &ndash; if someone wanted to write a C++ desktop app (or an iOS native app) that showed a cool introspection UI, they could.

Reply:
@tiziano88 tried the cfg variant, but it still fails to compile, both with the attribute and boolean variant of it. :/

Reply:
to recap, the two alternative approaches seem:

- bundle everything in the oak_loader binary itself, including static assets and a web server that can serve them. This has the advantage of keeping things self-contained in the binary, no additional dependencies when deploying, no need to multiplex multiple paths on the same host to make sure CORS headers line up, no need to worry about backwards compatibility of the interface vs the viewer (since they will always be built and deployed together), simpler way of testing that everything still works together
- only expose the API endpoint from oak_loader, but require a separate server for the static assets. The main advantage of this, AFAICT, is that it makes oak_loader less coupled with the static asset generation process, and that it requires less logic in the oak_loader itself (though note that the HTTP server needs to stay there for the API endpoint anyways)

@daviddrysdale @juliettepretot is this a fair summary? did I miss anything?

Reply:
That's an accurate summary. :) I have a slight preference for bundling the client for ease of use.

One minor thing I would add that CORS header are relatively straightforward in this case. Bundling a default client does not forbid opening the underlying API to outside developers now or in the future.

```rust
response
    .headers_mut()
    // Allow browsers from any origin to access this resource. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin
    .insert(ACCESS_CONTROL_ALLOW_ORIGIN, "*".parse().unwrap());
```

Reply:
My preference is also for the bundling approach, at least to begin with. If we find out that it's too complicated to make sure the assets make it through, then we can definitely reconsider and un-bundle them; it will be an extra data point that we will have gained in the meanwhile.

@daviddrysdale are you ok with going ahead with this?

Reply:
My preference is obviously for the loosely-coupled approach, but I guess I'm out-voted.  At least I won't have to install/use npm just to run Rust unit tests though&hellip;